archive/imo/imo1969_q1.lean

```diff
@@ -0,0 +1,81 @@
+import tactic
+import tactic.basic
+import tactic.linarith
+import tactic.norm_cast
+import tactic.ring
+
+open int
+open nat
+
+/-
+The 1969 IMO, problem 1:
+Prove that there are infinitely many natural numbers $a$ with the following property:
+the number $z = n^4 + a$ is not prime for any natural number $n$.
+
+The key to the solution is that you can factor this into the product of two polynomials, if a = 4*m^4.
+-/
+
+lemma factorization (m n: ℤ): (n^2 + 2*m^2 - 2*n*m) * (n^2 + 2*m^2 + 2*n*m) = n^4 + 4*m^4
```

Comment:
This is sometimes known as Sophie Germain's identity, and is useful for more than just this one problem. Maybe it belongs in mathlib proper (for any commutative ring) rather than in `archive`?

https://www.cut-the-knot.org/blue/SophieGermainIdentity.shtml

Reply:
See also https://en.wikipedia.org/wiki/Aurifeuillean_factorization

Reply:
My instinct would be that a theorem that is just proved with tactic.ring isn't too valuable to include in mathlib, but I don't feel strongly about it.

Reply:
mathlib has a great number of trivial lemmas with one-line proofs. The fact that this one has a name at least tends to suggest it might be appropriate for mathlib, with a doc string mentioning that name.

Reply:
Where do you think it should go?

Reply:
Probably in `algebra/group_power/basic.lean` (it's true for any commutative ring, and when such lemmas about rings involve `^` they seem to go there rather than under `algebra/ring/`).

Reply:
After looking it over, the lemmas in `basic.lean` seem really like basic, fundamental ones. Sophie Germain's identity is just like, a funny side effect of multiplying these particular polynomials together that's occasionally relevant for problem-solving. I don't think it belongs in there. Let's just leave this in the area specific to this IMO problem.

Reply:
I agree with @jsm28 here. We tend to err on the side of putting more in the main library rather than less.

Reply:
When I simply paste this code into `algebra/group_power/basic.lean`, it starts failing to resolve imports. Perhaps some issue with using tactic.ring inside the ring library itself? Let's just leave the core mathlib alteration out of scope of this pull request.

Reply:
Fair enough. I'll take care of it later then.